### PR TITLE
[sui framework] add `type_tag_bytes()` native function

### PR DESCRIPTION
**Pull request by leina05**
Native function to provide bytes of type tags. Useful for maintaining a registry of objects based on type. For example, to maintain a registry of AMM pools with type `Pool<X, Y>` one could store pool objects in a bag keyed on the pool's type tag.

### CODE REVIEW

File: crates/sui-framework/sources/types.move [hunk 1]
```diff
@@ -8,4 +8,6 @@ module sui::types {
     /// Tests if the argument type is a one-time witness, that is a type with only one instantiation
     /// across the entire code base.
     public native fun is_one_time_witness<T: drop>(_: &T): bool;
+
+    public native fun type_tag_bytes<T>(): vector<u8>;
```
Comment (awelc) [inline crates/sui-framework/sources/types.move:12]: It would be good to have a comment on what this function does

**Review by awelc — CHANGES_REQUESTED** [review]
This will break the Prover (I am currently trying to enable Prover runs in CI). It needs the following Boogie procedure added in the `Native Sui types implementation for object type `{{instance.suffix}}`` section of the "crates/sui/src/sui_move/sui-natives.bpl" file

```
procedure {:inline 1} $2_types_type_tag_bytes{{S}}(_: {{T}}) returns (res: Vec (int));
```

**Comment by tnowacki** [comment]
I would suggest two alternative approaches:

(1) 
If you are using bag, and you want a type based key, you could always do:
```
struct TypeKey<phantom T> has copy, drop, store { }
```
And then use this `TypeKey` in your bag. 
```
let bag: &mut Bag = ...;
Bag::add(bag, TypeKey<Foo> {}, Baz { ... })
Bag::add(bag, TypeKey<Bar> {}, Baz { ... })
```

(2)
If you want some actual bytes there for comparing the types, you can use `std::type_name`. Which can give you exactly what you want with bytes as described above, either in a structured form (with `type_name::TypeName<T>`), a string form (with `type_name::into_string` and `type_name::borrow_string`), or a byte form (with following it up with `std::ascii::as_bytes`)

**Comment by tnowacki** [comment]
Does approach (1) or (2) solve your problem, @leina05?

**Comment by leina05** [comment]
> Does approach (1) or (2) solve your problem, @leina05?

Yep, thanks so much! Was looking for exactly `type_name` and somehow missed it.

**Comment by tnowacki** [comment]
Glad to have helped! Closing for now then :)

**Pull request closed without merging**